products/llbuildSwift/Internals.swift

```diff
@@ -46,39 +46,39 @@ internal func stringFromData(_ data: llb_data_t) -> String {
 }
 
 extension Array where Element == String {
-    internal func withCArrayOfStrings<T>(_ body: @escaping (UnsafePointer<UnsafePointer<CChar>>) -> T) -> T {
-        func appendPointer(_ index: Array.Index, to target: inout Array<UnsafePointer<CChar>>) -> T {
-            if index == self.endIndex {
-                return body(&target)
-            } else {
-                return self[index].withCString { cStringPtr in
-                    target.append(cStringPtr)
-                    return appendPointer(self.index(after: index), to: &target)
-                }
+    private func withTemporaryBuffer<T>(_ body: @escaping (UnsafePointer<UnsafePointer<CChar>>) -> T) -> T {
+        // The buffer is in the form "a1\0a2\0a3\0…"
+        let totalLength = reduce(0) { $0 + $1.utf8.count + 1 }
+        let buffer = UnsafeMutableRawPointer.allocate(byteCount: totalLength, alignment: 1)
+        let elementPointers = UnsafeMutablePointer<UnsafePointer<CChar>>.allocate(capacity: count)
+        var bufferCursor = buffer
+        for (index, element) in enumerated() {
+            let elementPtr = element.withCString { ptr -> UnsafePointer<CChar> in
+                let result = UnsafePointer(bufferCursor.assumingMemoryBound(to: CChar.self))
+                bufferCursor.copyMemory(from: ptr, byteCount: element.utf8.count)
+                bufferCursor = bufferCursor.advanced(by: element.utf8.count)
+                bufferCursor.copyMemory(from: "\0", byteCount: 1)
+                bufferCursor = bufferCursor.advanced(by: 1)
+                return result
             }
+            elementPointers[index] = elementPtr
+        }
+        let result = body(UnsafePointer(elementPointers))
+        buffer.deallocate()
```

Comment:
Suggest moving this up to a `defer` right below buffer... less risky we leak if things are restructured later.

Or, can you use one of the `with` methods that takes a block?

Reply:
I added a `defer` after allocation of the pointers.